### ISSUE TITLE
STYLE fix pylint issues

### ISSUE THREAD

**Issue by MarcoGorelli** [pandas-dev/pandas#48855]
In https://github.com/pandas-dev/pandas/pull/48759/files we're introduced `pylint`, but have turned off its warnings as there's a _lot_ of them:

https://github.com/pandas-dev/pandas/blob/d719840e5a2639babab3b4646b11a111547c518f/pyproject.toml#L36-L67

Task here is:

1. pick one of the above warnings / errors, and remove it from `pyproject.toml`
2. run `pre-commit run pylint --hook-stage manual --all-files`. If this already passes, skip to step 4.
3. fixup any warnings that result
4. if `pre-commit run pylint --hook-stage manual --all-files` passes, then stage and commit your changes and open a pull request

If you believe a warning is a false-positive, then it's OK to ignore it in-line, e.g. 
```python
df.rename(id, mapper=id)  # pylint: disable=redundant-keyword-arg
```

Please comment here which pylint warning you'll work on before starting so we don't duplicate work. No need to ask for permission to work on this, and no need to comment "take" as multiple people can work on this concurrently

---

NOTE: please do _not_ write "closes #48855" in your PR

---

checks we probably can't turn off:
- too-many-function-args
- unsubscriptable-object
- unsupported-assignment-operation
- no-name-in-module
- no-member
- import-error
- not-an-iterable
- invalid-unary-operator-type
- invalid-name
- wrong-import-order (and anything else related with imports)
- unexpected-keyword-argument
- use-implicit-booleaness-not-len
- use-implicit-booleaness-not-comparison
- comparison-with-itself
- no-else-*
- line-too-long
- overridden-final-method
- singleton-comparison
- unsupported-membership-test
- pointless-statement
- broad-except
- undefined-variable

If you're not sure whether a check should fit into the above list, please do ask

**Comment by shivamurali** [comment]
Hi there! I'll get to work on the "used-before-assignment" warning and remove it from the pyproject.toml file.

**Comment by Exoutia** [comment]
Hi i am new to open source i want to contribute to this project by fixing this but i cant seem to run pylint with --all-files option am i doing something wrong.
or do i need to do thing before hand beside making a branch and cloning the repo can someone help me

**Comment by MarcoGorelli** [comment]
Hi @Exoutia - please show your output

The command is `pre-commit run pylint --all-files`

**Comment by Exoutia** [comment]
> Hi @Exoutia - please show your output
> 
> The command is `pre-commit run pylint --all-files`

Thanks, now it ran some problems has happened during installation of pre-commit i again installed it and it got fixed
![image](https://user-images.githubusercontent.com/94464106/193232132-08cc213b-f93d-485b-bdba-cd6f8f1c1a3a.png)

**Comment by slackline** [comment]
I'll work on removing `import-error`.

**Comment by gulyapulya** [comment]
Hi! I am also new to the open source, but I want to contribute, I will work on the "no-method-argument".

**Comment by kostyafarber** [comment]
Hi! I'll work on `inherit-non-class`.

**Comment by soumilbaldota** [comment]
Hi, I will take `bad-super-call` warning, I have added the pull request above.

**Comment by FloHofstetter** [comment]
Hi, I will take `function-redefined` warning.

**Comment by rokaicker** [comment]
Hi ! I'm new to open source, but I'd like to tackle the `too-many-function-args`!

**Comment by MarcoGorelli** [comment]
thanks @rokaicker - that one's probably too hard to turn off for now, without going through a deprecation cycle, but there's lot's of others to choose

I've added this one to the description as one that's probably not possible to turn off for now

**Comment by jetale** [comment]
working on  "unsupported-assignment-operation"

**Comment by Exoutia** [comment]
i am working on "no-self-argument"

**Comment by shalearkane** [comment]
I am working on "unexpected-keyword-arg"

**Comment by roadswitcher** [comment]
I am working on `no-name-in-module`

**Comment by Leviob** [comment]
Hi, I would like to work on "used-before-assignment," but I'm aware it was already claimed by @shivamurali. Can I submit my own PR for it? Thanks.

**Comment by joaomarcoscrs** [comment]
I'll work on the `not-callable`

**Comment by rAwsam** [comment]
Hi! I am also new to the open source, but I want to contribute, I will work on the "undefined-variable"

**Comment by TejasCreative** [comment]
I'll take `no-member`

**Comment by sakalmon** [comment]
I'll work on `unsubscriptable-object`

**Comment by sakalmon** [comment]
> I'll work on `unsubscriptable-object`

I'm new to open source but I removed the above line and got numerous errors. All of them were "EE1136 - unsubscriptable (unsubscriptable-object)". 

[Pylint documentation](https://pylint.pycqa.org/en/latest/user_guide/messages/error/unsubscriptable-object.html) suggests it may be because there is no `__get__item` method or `__class_getitem__` for the class. I believe this is a false positive and the error will have to remain disabled.

**Comment by MarcoGorelli** [comment]
thanks for looking into this @sakalmon - that's fine, I'll add that to the list of ones we can't turn off

**Comment by sakalmon** [comment]
@MarcoGorelli No worries, happy to help!

**Comment by roadswitcher** [comment]
> I am working on `no-name-in-module`

Disabling this resulted in a substantial number of E0611s that are tied to referencing calls in underlying C implementation within the `pandas._libs/*` directories.   [Pylint documentation for other errors](https://pylint.pycqa.org/en/stable/user_guide/messages/error/no-member.html) implies that there are flags that could be invoked in order to make pylint start building AST trees from the underlying C modules, but that's probably out of scope for this issue.

 As with @sakalmon's investigation of `unscriptable-object`, I think this should remain disabled for now.

**Comment by TejasCreative** [comment]
> I'll work on `no-member`!

Looks like no-member can't be turned off. After removing it, A wide variety of E1101: Instance of xyz has no abc member. Looks like this is a false positive for object members that are created dynamically, but exist at the time they are accessed.

**Comment by techrajdeep** [comment]
I am looking for an issue for my first contribution . Can you help me here.

**Comment by slackline** [comment]
> I'll work on removing `import-error`

As mentioned above I've opened PR #49013 which hopefully removes the need to disable `import-error`.

**Comment by slackline** [comment]
I was trying to work out what had been taken and what was still available, knocked up this table and thought it might be useful.

| Pylint Error                       | Worked on                  | PR              | Comment                                                                                                   |
|:-----------------------------------|:--------------------------:|:----------------|:----------------------------------------------------------------------------------------------------------|
| `C`                                | [ ]                        |                 |                                                                                                           |
| `R`                                | [ ]                        |                 |                                                                                                           |
| `W`                                | [ ]                        |                 |                                                                                                           |
| `abstract-class-instantiated`      | [ ]                        |                 |                                                                                                           |
| `access-member-before-definition`  | [ ]                        |                 |                                                                                                           |
| `bad-super-call`                   | [x] @soumilbaldota         | #48896          |                                                                                                           |
| `c-extension-no-member`            | [ ]                        |                 |                                                                                                           |
| `function-redefined`               | [x] @FloHofstetter         | #48906          |                                                                                                           |
| `import-error`                     | [x] @slackline             | #49013          | Can't be removed see PR discussion.                                                                       |
| `inherit-non-class`                | [x] @kostyafarber          | #48897          |                                                                                                           |
| `invalid-repr-returned`            | [ ]                        |                 |                                                                                                           |
| `invalid-unary-operand-type`       | [ ]                        |                 |                                                                                                           |
| `misplaced-bare-raise`             | [ ]                        |                 |                                                                                                           |
| `no-member`                        | [x] @TejasCreative         |                 | Can't be removed see [comment](https://github.com/pandas-dev/pandas/issues/48855#issuecomment-1272142781) |
| `no-method-argument`               | [x] @gulyapulya            |                 |                                                                                                           |
| `no-name-in-module`                | [x] @roadswitcher          |                 | Can't be removed see [commen](https://github.com/pandas-dev/pandas/issues/48855#issuecomment-1271503075)  |
| `no-self-argument`                 | [x] @Exoutia               | #48912          |                                                                                                           |
| `no-value-for-parameter`           | [x] @shalearkane           | #48917          |                                                                                                           |
| `non-iterator-returned`            | [ ]                        |                 |                                                                                                           |
| `not-an-iterable`                  | [ ]                        |                 |                                                                                                           |
| `not-callable`                     | [x] @joaomarcoscrs         |                 |                                                                                                           |
| `redundant-keyword-arg`            | [ ]                        |                 |                                                                                                           |
| `too-many-function-args`           | [ ]                        |                 | Not possible to turn off at present.                                                                      |
| `undefined-variable`               | [x] @rAwsam                |                 |                                                                                                           |
| `unexpected-keyword-arg`           | [x] @shalearkane           |                 |                                                                                                           |
| `unpacking-non-sequence`           | [ ]                        |                 |                                                                                                           |
| `unsubscriptable-object`           | [x] @sakalmon              |                 | Can't be removed see [comment](https://github.com/pandas-dev/pandas/issues/48855#issuecomment-1269441871) |
| `unsupported-assignment-operation` | [x] @jetale                | #48982          |                                                                                                           |
| `unsupported-membership-test`      | [ ]                        |                 |                                                                                                           |
| `used-before-assignment`           | [x] @shivamurali / @Leviob | #48895 / #48940 |                                                                                                           |

**Comment by techrajdeep** [comment]
@slackline , I am looking for an issue for my first contribution , is there anything available here. Can you please help me

**Comment by slackline** [comment]
@techrajdeep rows in the above table are pylint errors that have not been worked on yet. See the original post in this issue for how to go about investigating the source of the errors. If you're not familiar with the concept of linting and/or pylint then it would be worth checking the [documentation](https://pylint.pycqa.org/en/latest/).

**Comment by vamsi-verma-s** [comment]
working on `invalid-repr-returned`

**Comment by roadswitcher** [comment]
Working on `not-an-iterable`

**Comment by roadswitcher** [comment]
Working on `non-iterator-returned` now

**Comment by roadswitcher** [comment]
Checking `c-extension-no-member` now

**Comment by roadswitcher** [comment]
> Checking c-extension-no-member now

OK -- pylint found no instances of this in the code on my local machines, but if I enable pylint checking in the build by deleting this line from `pyproject.toml`, it fails one of the build processes in the Ubuntu workflow?     Interesting.

**Comment by roadswitcher** [comment]
Checking 'C' now.

**Comment by roadswitcher** [comment]
> Checking `C` now.

1323 separate files, >40k identified problems.    I suggest that check is a candidate to leave disabled.
```
(pandas-dev) root@007e63d6aeef:/home/pandas# pre-commit run pylint --all-files | tee pylint-output
(pandas-dev) root@007e63d6aeef:/home/pandas# cat pylint_output | grep " Module" | wc -l
1323
(pandas-dev) root@007e63d6aeef:/home/pandas# cat pylint_output | wc -l
44157
```
I'm checking `R` and `W` now.

**Comment by MarcoGorelli** [comment]
thanks @roadswitcher for checking!

C, R, W are codes which have a lot of codes behind them. A PR to replace `'C'` in `pyproject.toml` with all the individual codes would be very welcome (likewise for R and W)

The code I'm particularly keen on enabling is `W0621`, as that one would've helped prevent a bug in pandas 1.5.0

**Comment by jpfa1406** [comment]
I'll work in `redundant-keyword-arg`

**Comment by ayushanand308** [comment]
HI! I am a first time contributor. Can anyone assign me a beginner friendly issue so that I can work on it.

**Comment by MarcoGorelli** [comment]
you can work on this issue without being assigned it. if you want to work on a different issue, you can comment there

**Comment by vamsi-verma-s** [comment]
All `abstract-class-instantiated` errors are because of `pandas.ExcelWriter` . 
https://github.com/pandas-dev/pandas/blob/e3d0126e8af9f0a23dd6609a23817f858825e089/pandas/io/excel/_base.py#L1090-L1110

This is recommended way to create writer objects. Should this be considered acceptable and disable this error?

>To write a single object to an Excel .xlsx file it is only necessary to specify a target file name. To write to multiple sheets it is necessary to create an ExcelWriter object with a target file name, and specify a sheet in the file to write to.

**Comment by vamsi-verma-s** [comment]
worked on `access-member-before-definition`

**Comment by roadswitcher** [comment]
I took a look at enabling the  `invalid-unary-operator-type` check, and I think it's exposed a problem where pylint doesn't properly recognize some NumPy constructs.

Probably should leave the check disabled for now, but I'm working on a minimal test case to submit as an issue to the Pylint project.

**Comment by PedroPUCRIO** [comment]
HI! I am a first time contributor. Can I check unpacking-non-sequence?

**Comment by Leviob** [comment]
I'm working on expanding the "C", "R", and "W" codes into their respective errors. I'll keep each of these grouped together in `pyproject.toml` so it is more clear which lines are disabling errors vs conventions, refactors, or warnings.

**Comment by Leviob** [comment]
After expanding the "C", "R", and "W" message types into their individual message names, here are the new pylint messages that can be worked on. 
https://github.com/pandas-dev/pandas/blob/d2acdee72b33c8c52bb5f8b902190bbfcf27ca47/pyproject.toml#L81-L204

I'm going to work on `superfluous-parens` next.

**Comment by Leviob** [comment]
Pylint produces nearly 20k messages for [invalid-name](https://pylint.pycqa.org/en/latest/user_guide/messages/convention/invalid-name.html). Most of these are due to variables named with less than 3 characters, and not necessarily violating naming rules. Unless pylint is changed to accept short variable names, or we want to rename every `df`, we probably can't enable it. 
See pylint issue: https://github.com/PyCQA/pylint/issues/2018

**Comment by fabinca** [comment]
I'm working on the following R-Type pylint messages
#  "literal-comparison",
#  "no-else-break",
#  "no-else-continue",
#  "no-else-raise",
#  "no-else-return",
#  "no-self-use",
#  "simplifiable-if-expression",
#  "simplifiable-if-statement",
#  "useless-option-value",
#  "useless-return",

**Comment by natmokval** [comment]
Hi! I'm working on the “wrong-import-order” C-Type pylint message.

**Comment by stellalin7** [comment]
I can work on the "disallowed-name" C-type pylint message

**Comment by manny-uncharted** [comment]
take

**Comment by MarcoGorelli** [comment]
no need to comment "take" as multiple people can work on this concurrently

**Comment by natmokval** [comment]
PR for “wrong-import-order”  is created: https://github.com/pandas-dev/pandas/pull/49390

**Comment by natmokval** [comment]
I'll work on the "useless-return" message.

**Comment by natmokval** [comment]
I'm going to work on the "nan-comparison" pylint message.

**Comment by stellalin7** [comment]
> I can work on the "disallowed-name" C-type pylint message
[issue 48855 enable pylint C-type "disallowed-name " warning #49379](https://github.com/pandas-dev/pandas/pull/49379)

**Comment by natmokval** [comment]
I'm going to work on the type "W" pylint warnings: `wildcard-import`, `unused-wildcard-import`.

**Comment by roadswitcher** [comment]
Working on W warnings `broad-except`, and `cell-var-from-loop`

**Comment by natmokval** [comment]
I'm working on the following R-Type pylint messages: `literal-comparison`, `no-self-use`, and `invalid-sequence-index`.

**Comment by natmokval** [comment]
Hi! I’ll work on `consider-using-sys-exit`.

**Comment by natmokval** [comment]
I'm going to work on the type "R" pylint warning: `comparison-of-constants`

**Comment by natmokval** [comment]
I’ll work on the type "W" pylint warning: `duplicate-value`.

**Comment by MarcoGorelli** [comment]
Note for anyone working on this: `pylint` is now run only in CI, so you'll need to run the command as
```
pre-commit run pylint --hook-stages manual --all-files
```

**Comment by natmokval** [comment]
> Note for anyone working on this: `pylint` is now run only in CI, so you'll need to run the command as
> 
> ```
> pre-commit run pylint --hook-stages manual --all-files
> ```

Hi @MarcoGorelli . I noticed a typo in the above command. The following command works for me 
```
pre-commit run pylint --hook-stage manual --all-files
```

**Comment by MarcoGorelli** [comment]
you're right, thanks @natmokval !

**Comment by uzzell** [comment]
I'll work on the type "W" warning `import-self`.

**Comment by MarcoGorelli** [comment]
I've opened a separate issue for `redefined-outer-name`, in case anyone here wants to contribute towards that one: https://github.com/pandas-dev/pandas/issues/49656

It'll take a while to enable it, but I really think it'll be worth it

**Comment by natmokval** [comment]
I'm working on `self-assigning-variable`

**Comment by uzzell** [comment]
I looked at `comparison-with-itself`.  If I counted correctly, there were 72 warnings, all of which were false positives.  Does it make sense to keep it enabled?

**Comment by MarcoGorelli** [comment]
Thanks @uzzell , yeah let's keep that one off thanks

**Comment by seanjedi** [comment]
I'm working on `wrong-import-order`

**Comment by MarcoGorelli** [comment]
thanks but we're not doing that one, check the end of the issue description

**Comment by Leviob** [comment]
I'll work on `undefined-loop-variable`.
I'll start working on `no-else-return` within the pandas/io directory.

**Comment by MarcoGorelli** [comment]
thanks @Leviob - some people aren't keen on `no-else-*` so let's leave that one out for now

**Comment by uzzell** [comment]
I'll work on `useless-parent-delegation`.

**Comment by JasmandeepKaur** [comment]
I'll work on `line-too-long`

**Comment by MarcoGorelli** [comment]
hi @JasmandeepKaur - that one's already covered by flake8, no need

**Comment by JasmandeepKaur** [comment]
Hello, Thanks for letting g me know! I'll look for another

**Comment by angularOcean** [comment]
I wanted to work on  `unused-argument`,  `unused-import`,  and` unused-variable`, but I got an error when trying to run `pre-commit run pylint --all-files`. I get the message "No hook with id 'pylint' in stage 'commit'", anyone else run into this issue? I forked the repo, made a branch and am running it on the branch as outline in the developer guide.

**Comment by MarcoGorelli** [comment]
you need `--hook-stage=manual` too

let's not touch anything related with imports

**Comment by angularOcean** [comment]
Thanks, I will leave `unused-import` alone. I tried `pre-commit run pylint --hook-stages manual --all-files` but now I get this error: "pre-commit: error: unrecognized arguments: --hook-stages manual"

**Comment by MarcoGorelli** [comment]
sorry, `--hook-stage`

**Comment by angularOcean** [comment]
Thanks that fixed the issue. Also, after running a check on `unused-argument`, there turned up several hundred lines with the issue and from a quick look of a sampling of them, a large proportion of them are false positives i.e. 'col' in:
`    def time_iteritems_indexing(self):
        for col in self.df3:
            self.df3[col]`
I can mark this as a false positive but there would be a lot of them. Would it be better to continue keeping warning turned off?

**Comment by MarcoGorelli** [comment]
we can probably keep it off

We're  probably scraping the barrel here with this issue, I think we've probably enabled most of the most meaningful ones 

There's still some files to do in https://github.com/pandas-dev/pandas/issues/49656 if you fancy that though

**Comment by seanjedi** [comment]
Is anyone working on `unsupported-membership-test`? 
Can I take that task?

**Comment by angularOcean** [comment]
I took a look and I think `use-implicit-booleaness-not-comparison` is worth working on too

**Comment by angularOcean** [comment]
I will also work on `use-implicit-booleaness-not-len` and see if that can enabled as well

**Comment by natmokval** [comment]
Hi @MarcoGorelli , I checked pylint warning `singleton-comparison` and suggest keeping it off. Only false-positive warnings are raised on 8 lines (in 4 files); on every line, flake8 warnings, such as noqa: E711 or noqa: E712, are ignored. 

For example:
https://github.com/pandas-dev/pandas/blob/main/pandas/tests/series/test_arithmetic.py#L827

https://github.com/pandas-dev/pandas/blob/main/pandas/tests/frame/indexing/test_where.py#L828

**Comment by seanjedi** [comment]
I've started to work on `unsupported-membership-test` if that is alright.
I started working on it here: https://github.com/pandas-dev/pandas/pull/49848
I have not opened up the PR for review yet since there are still some issues I am trying to debug, once I have them resolved then I will open up the PR.

**Comment by MarcoGorelli** [comment]
thanks @seanjedi - let's keep that one off too, I think they're all false positives and we risk annoying contributors

**Comment by CerqueiraMatheus** [comment]
I will work on `access-member-before-definition`

**Comment by seanjedi** [comment]
Alright, thanks for letting me know! 
I will work on `pointless-string-statement` then.

**Comment by seanjedi** [comment]
@MarcoGorelli I have worked on implementing the changes for `pointless-string-statement` here: https://github.com/pandas-dev/pandas/pull/49880

**Comment by roadswitcher** [comment]
Found just over 30 instances of  `broad-except` which appeared to be false positives -- do we want to keep this warning disabled??

**Comment by uzzell** [comment]
I'll take `unnecessary-list-index-lookup`.

**Comment by MarcoGorelli** [comment]
> Found just over 30 instances of `broad-except` which appeared to be false positives -- do we want to keep this warning disabled?? (ref PR #49910 for example)

yeah let's do that

**Comment by roadswitcher** [comment]
> > Found just over 30 instances of `broad-except` which appeared to be false positives -- do we want to keep this warning disabled?? (ref PR #49910 for example)
> 
> yeah let's do that

PR closed.

**Comment by bang128** [comment]
Has anyone taken line-too-long yet? Can I work on it?

**Comment by MarcoGorelli** [comment]
Thanks everyone for your contributions!

I think we've probably taken the most valuable ones already, so I think we can stop asking for contributions on this one.

If anyone comes across a pylint code which can be removed and which adds then value, then please do submit a PR to fix it! You're still welcome to do that, I just think it's time to close the issue

Thanks all 🙏
